### PR TITLE
Removed default World and Field Constructors

### LINKED ISSUE #1401: Remove isValid() from Field and disallow the creation of degenerate fields
### Description of the task

#1251 added the `isValid()` function to `Field` to check for zero-length fields because we did not want to publish `Worlds` without a valid field.

#1385 then removed the default constructor from `Field` because there were several instances where we were just manually "default constructing" fields with all zeroes.

What this has led to is the easy ability to create degenerate fields that cause later problems for consumers. It's an anti-pattern to allow invalid Filds to be created, and provide an `isValid()` function that we _trust_ people to check before doing anything with a field.

To remedy this, we should
* Delete the default constructor for `Field`
* Remove the `isValid()` function
* Throw an exception if a field is created with any zero-length parameters

Any code that was default-constructing fields unnecessarily is likely a smell that something is being given more data than it needs (eg. being given the World when all it needs is the ball / robots).

### Acceptance criteria

- [ ] Delete the default constructor for `Field`
- [ ] Remove the `isValid()` function
- [ ] Throw an exception if a field is created with any zero-length parameters

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
* Removed `World` constructor
* Removed `Field` constructor
* Make everything still work
* Removed `Field::isValid`
* Removed `RobotExpiryBufferMilliseconds` dynamic param

### Testing Done
CI

### Resolved Issues
resolves #1401 
part of #1361 

### Length Justification
Not a lot of new code

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Comment by MathewMacDougall** [comment]
I'd like to try avoid optional worlds all around the coroutines of STP. Will try poke around with this locally when I have time :tm:

File: src/software/ai/hl/stp/stp_refbox_game_state_play_selection_test.cpp [hunk 1]
```diff
@@ -34,11 +34,11 @@ class STPRefboxGameStatePlaySelectionTestWithPositions
    protected:
     void SetUp() override
     {
-        world.mutableField() = ::Test::TestUtil::createSSLDivBField();
+        world.updateField(::Test::TestUtil::createSSLDivBField());
```
Comment (garethellis0) [inline src/software/ai/hl/stp/stp_refbox_game_state_play_selection_test.cpp]: I think `setField` might make more sense? `update` suggests the potential for a partial change of some type (at least to me)

File: src/software/sensor_fusion/sensor_fusion.h [hunk 2]
```diff
@@ -40,17 +40,24 @@ class SensorFusion : public Subject<World>, public ThreadedObserver<SensorMsg>
...
+     * @param new data
+     */
+    void updateWorldComponents(const SSL_WrapperPacket &packet);
+    void updateWorldComponents(const Referee &packet);
+    void updateWorldComponents(
         const google::protobuf::RepeatedPtrField<TbotsRobotMsg> &tbots_robot_msgs);
-    void updateWorld(const SSL_GeometryData &geometry_packet);
-    void updateWorld(const SSL_DetectionFrame &ssl_detection_frame);
+    void updateWorldComponents(const SSL_GeometryData &geometry_packet);
+    void updateWorldComponents(const SSL_DetectionFrame &ssl_detection_frame);
```
Comment (garethellis0) [inline src/software/sensor_fusion/sensor_fusion.h]: Hmm, I think I prefer just updateWorld, I don't think `Components` really adds a ton of value here?

File: src/software/world/ball.h [hunk 3]
```diff
@@ -11,6 +11,8 @@
 class Ball final
 {
    public:
+    Ball();
```
Comment (garethellis0) [inline src/software/world/ball.h]: Why did we add a default `Ball` constructor while removing those for `Field` and `World`?

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Looking good! Really happy to see this.

File: src/software/world/ball.h (hunk 3 above)
Comment (jonathanlew) [reply src/software/world/ball.h]: it seemed like bit of work to remove initially, but it's not so bad in the end

File: src/software/sensor_fusion/sensor_fusion.h (hunk 2 above)
Comment (jonathanlew) [reply src/software/sensor_fusion/sensor_fusion.h]: removed

File: src/software/ai/hl/stp/stp_refbox_game_state_play_selection_test.cpp (hunk 1 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/stp_refbox_game_state_play_selection_test.cpp]: removed it completely since it was only used in tests

**Comment by jonathanlew** [comment]
> I'd like to try avoid optional worlds all around the coroutines of STP. Will try poke around with this locally when I have time tm

i would like to merge this before #1418 and other work that will use World, Field, etc.

**Comment by MathewMacDougall** [comment]
> i would like to merge this before #1418 and other work that will use World, Field, etc.

oof. Ok I will try prioritize looking at this.

**Comment by MathewMacDougall** [comment]
Check out #1428 for a quick example of the coroutine change

**Review by garethellis0 — DISMISSED** [review]
LGTM :+1: :tada:

**Comment by jonathanlew** [comment]
@MathewMacDougall Made the change to pass World into getNextTactic

**Review by EvanMorcom — APPROVED** [review]
LGTM 🏑

**Review by MathewMacDougall — APPROVED** [review]
LGTM. Definitely made me notice lots of places we can use more granular datastructures, but that's an entirely separate issue. Nice work with this :+1:

**Review by akhilveeraghanta — APPROVED** [review]
lgtm 🌎

**Pull request merged**